### PR TITLE
Add RepeatVector layer

### PR DESCRIPTION
**Pull request by dosier**

### CODE REVIEW

**Review by dosier — COMMENTED** [review]
This commit can be ignored. It is reverted later during a merge with master.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ * Input shape: `2D tensor of shape (num_samples, features)`.
+ *
+ * Output shape: `3D tensor of shape (num_samples, n, features)`.
+ *
+ * @property n Repetition factor.
+ * @property [name] Custom layer name.
+ * @constructor Creates [RepeatVector] object.
+ *
+ * @author Stan van der Bend
+ * @since 0.2
```
Comment (avan1235) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:28]: I suppose you mean 0.3

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ */
+public class RepeatVector(
+    public val n: Int,
+    name: String = ""
+) : Layer(name) {
+
+    override fun build(tf: Ops, kGraph: KGraph, inputShape: Shape) {
```
Comment (avan1235) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:35]: I propose
```kotlin
override fun build(tf: Ops, kGraph: KGraph, inputShape: Shape) = Unit
```
as this approach describes itself and doesn't need any comments that brings nothing to reader (as Kotlin gives us the bility to write such code, not like in Java)

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        //left empty
+    }
+
+    override fun computeOutputShape(inputShape: Shape): Shape {
+        require(inputShape.numDimensions() == 2) { "input tensor must have 2 dimensions" }
```
Comment (avan1235) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:40]: I saw only the phrases starting from big letter as checks reports. Keep the convention to have a well documented library

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        val tensorShape = TensorShape(inputShape)
+        // TODO: maybe make `n` of type Long?
```
Comment (avan1235) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:42]: imo you should as for now most of the ocnfiguration properties seems to have Long type.
btw why not just 
```kotlin
return Shape.make(inputShape[0], n, inputShape[1])
```
the `tensorShape` is unnecessary.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        return Shape.make(tensorShape[0], n.toLong(), tensorShape[1])
+    }
+
+    override fun forward(
+        tf: Ops,
+        input: Operand<Float>,
+        isTraining: Operand<Boolean>,
+        numberOfLosses: Operand<Float>?
+    ): Operand<Float> {
+        return tf.repeat(input, n)
```
Comment (avan1235) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:52]: there are no benefits from creating the `Ops.repeat` as separate function (it's not going to be reused anywhere). Move your implementation to only `forward` as it is short enough

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        set(value) = assignWeights(value)
+
+    override val hasActivation: Boolean get() = false
+
+    override val paramCount: Int get() = 0
+
+    override fun toString(): String {
+        return "RepeatVector"
+    }
+}
```
Comment (avan1235) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:72]: you can fix this missing newline with other changes

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2020 JetBrains s.r.o. and Kotlin Deep Learning project contributors. All Rights Reserved.
```
Comment (avan1235) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:2]: Once again 2021 😉

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import org.jetbrains.kotlinx.dl.api.core.KGraph
+import org.jetbrains.kotlinx.dl.api.core.layer.reshaping.RepeatVector
+import org.jetbrains.kotlinx.dl.api.core.shape.toIntArray
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.Test
+import org.tensorflow.Graph
+import org.tensorflow.Output
+import org.tensorflow.op.Ops
+
+/**
```
Comment (avan1235) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:17]: I see no comments in tests classes. Keep the convention imo to have clean code in project

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ * @property [name] Custom layer name.
+ * @constructor Creates [RepeatVector] object.
+ *
+ * @author Stan van der Bend
+ * @since 0.2
+ */
+public class RepeatVector(
+    public val n: Int,
+    name: String = ""
+) : Layer(name) {
```
Comment (avan1235) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:33]: you should check in init block if `n >= 1` like 
```kotlin
init {
    require(n >= 1) { "Number of repetitions (n) in RepeatVector should be positive but got $n" }
}
```

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+ * A test for the [RepeatVector] layer.
+ *
+ * @author Stan van der Bend
+ */
+internal class RepeatVectorLayerTest {
+
+    @Test
+    fun `test output shape`(){
```
Comment (avan1235) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:25]: file needs reformatting (auto format in Intellij with Ctrl+Alt+L)

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        val actual = y.tensor().copyTo(Array(3) { Array(layer.n) { FloatArray(3) } })
+        val expected = arrayOf(
+            arrayOf(floatArrayOf(0F, 1F, 2F), floatArrayOf(0F, 1F, 2F)),
+            arrayOf(floatArrayOf(0F, 1F, 2F), floatArrayOf(0F, 1F, 2F)),
+            arrayOf(floatArrayOf(0F, 1F, 2F), floatArrayOf(0F, 1F, 2F))
+        )
+        Assertions.assertArrayEquals(expected, actual)
+    }
+
+    private operator fun RepeatVector.invoke(input : Array<FloatArray>) : Output<Float> {
```
Comment (avan1235) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:46]: Very good idea from my perspective but maybe open a separate issue for this or join [this discussion](https://github.com/JetBrains/KotlinDL/discussions/132) with your improvement to test readability and some proposal. Btw it should be done for the `Layer` and then reused in all tests in the way you did this.

Comment (avan1235) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:46]: btw clean this code a little bit (I don't see the need to refer to `Shape` as `org.tensorflow.Shape`) and let's see that the block with single value `{ return value }` can be reduced to `= value` (especially when you use the `.let { }`)

**Review by avan1235 — CHANGES_REQUESTED** [review]
I leave m thoughts, don't worry about them a lot but try to get from them as much as you can as in my opinion this code needs most of these fixes but by doing this you can learn creating cleaner code

**Comment by avan1235** [comment]
> Commits [78196a8](https://github.com/JetBrains/KotlinDL/commit/78196a88018bad65df0adbf925ffc52862bd6abf) and [6347d65](https://github.com/JetBrains/KotlinDL/commit/6347d656d898fde3d2084e9b9489f6cddf1ed150) can be ignored. They are reverted in [208cd56](https://github.com/JetBrains/KotlinDL/commit/208cd56a913a2c85e6bafe96a70e51aef83d5d71)

Just FYI your commits are squashed before merge so I think the reviewer looks only on the final version of commit so the steps that you did are not so important - the final result counts only 😉

**Comment by dosier** [comment]
@avan1235 thanks a lot for your comments. This is my first time contributing to an open source project and your comments were really thoughtful, will make changes and keep in mind for future contributions.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt (hunk 3 above)
Comment (dosier) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:40]: I am a bit confused on this point. I wasn't sure on how to handle this assertion, so I used something I saw used elsewhere in the project. Any suggested change?

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt (hunk 2 above)
Comment (dosier) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:35]: @avan1235 Nice! I never thought of that. I copied that piece from another similar Layer class. We might wanna convert all of the cases to this. It's a lot cleaner.

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt (hunk 4 above)
Comment (dosier) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:42]: `Shape` doesn't have operator functions for get, I could do:
```Kotlin
        val input = inputShape.toLongArray()
        return Shape.make(input[0], n.toLong(), input[1])
```

Comment (mkaze) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:44]: @dosier You can simply use `size`:
```kotlin
return Shape.make(inputShape.size(0), n.toLong(), inputShape.size(1))
```

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt (hunk 3 above)
Comment (avan1235) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:42]: I was only thinking about a small typo here so just write 
```kotlin
require(inputShape.numDimensions() == 2) { "Input tensor must have 2 dimensions but got ${inputShape.numDimensions()}" }
```
as I understand one of KotlinDL goals i sto provide readable errors for user that can be easily fixed

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt (hunk 4 above)
Comment (dosier) [reply api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/reshaping/RepeatVector.kt:44]: @mkaze ah of course, thanks

**Comment by zaleslaw** [comment]
Hi, @dosier is it ready for review?

File: api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelSaver.kt [hunk 12]
```diff
@@ -530,6 +532,16 @@ private fun createKerasFlattenLayer(layer: Flatten): KerasLayer {
     return KerasLayer(class_name = LAYER_FLATTEN, config = configX)
 }
 
+private fun createKerasRepeatVectorLayer(layer: RepeatVector): KerasLayer {
+    val configX = LayerConfig(
```
Comment (zaleslaw) [inline api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/inference/keras/ModelSaver.kt:536]: Don't forget trainable status

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import org.junit.jupiter.api.Test
+import org.tensorflow.Graph
+import org.tensorflow.Output
+import org.tensorflow.Shape
+import org.tensorflow.op.Ops
+
+internal class RepeatVectorLayerTest {
+
+    @Test
+    fun `test output shape`() {
```
Comment (zaleslaw) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:21]: please make it more consistent, rename it without ``

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+        val expected = arrayOf(
+            arrayOf(floatArrayOf(0F, 1F, 2F), floatArrayOf(0F, 1F, 2F)),
+            arrayOf(floatArrayOf(0F, 1F, 2F), floatArrayOf(0F, 1F, 2F)),
+            arrayOf(floatArrayOf(0F, 1F, 2F), floatArrayOf(0F, 1F, 2F))
+        )
+        Assertions.assertArrayEquals(expected, actual)
+    }
+
+    // TODO: generalise this for Layer, see https://github.com/JetBrains/KotlinDL/issues/145
+    private operator fun RepeatVector.invoke(input: Array<FloatArray>): Output<Float> = Ops.create().let { tf ->
```
Comment (zaleslaw) [inline api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:43]: maybe you could make a util internal extension function for the Layer class?

**Review by zaleslaw — CHANGES_REQUESTED** [review]
Please fix a few minor things

File: api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt (hunk 14 above)
Comment (dosier) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:43]: maybe I could, u mean as described in #145 ? 
What should I do for now though, remove the function?

Comment (zaleslaw) [reply api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/RepeatVectorLayerTest.kt:50]: So at this moment, you could leave it as is, I suppose.
The improvement related to #145 should be implemented in the separate PR and discussed with other contributors faced with this problem.

**Pull request merged**